File: Survey_Basket.UI/src/app/core/models/service-result.ts
export interface ServiceError {
  description: string;
  code: number;
}

export interface ServiceResult<T> {
  succeeded: boolean;
  data?: T;
  error?: ServiceError;
}

export interface ServiceListResult<TItem, TStats> {
  items: PagedList<TItem>;
  stats: TStats;
}

export interface PagedList<T> {
  items: T[];
  pageNumber: number;
  totalPages: number;
  totalCount: number;
  hasPreviousPage: boolean;
  hasNextPage: boolean;
}

File: Survey_Basket.Api/Models/ServiceResult.cs
namespace Survey_Basket.Api.Models;

public sealed record ServiceError(string Description, int Code);

public sealed class ServiceResult<T>
{
    public bool Succeeded { get; init; }
    public T? Data { get; init; }
    public ServiceError? Error { get; init; }

    public static ServiceResult<T> Success(T data) => new()
    {
        Succeeded = true,
        Data = data
    };

    public static ServiceResult<T> Failed(ServiceError error) => new()
    {
        Succeeded = false,
        Error = error
    };
}
